streams/src/test/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdaterTest.java

```diff
@@ -1043,6 +1046,28 @@ public void shouldResumeActiveStatefulTask() throws Exception {
         verify(changelogReader, times(2)).enforceRestoreActive();
     }
 
+    @Test
+    public void shouldAwaitWhenAllTasksPaused() throws Exception {
+        final StreamTask task = statefulTask(TASK_0_0, mkSet(TOPIC_PARTITION_A_0)).inState(State.RESTORING).build();
+        stateUpdater.start();
+        stateUpdater.add(task);
+
+        when(topologyMetadata.isPaused(null)).thenReturn(true);
+
+        verifyPausedTasks(task);
+
+        reset(changelogReader);
+        Thread.sleep(100);
+        verify(changelogReader, never()).restore(any());
```

Comment:
Instead of this, we could expose the thread state on the implementation of the state updater, but not on the interface. With that and a wait until verification, we could verify that the thread changes to the waiting state. We similar with `getActiveTasks()` or `getPausedTasks()`. That does also smell a bit, but it avoids the sleep and maybe smells a little bit less? 🙂

Reply:
I like the idea!